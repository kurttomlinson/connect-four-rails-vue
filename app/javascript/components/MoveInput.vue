<template>
  <div class="board-container">
    <span v-for="(valid, column) in validMoves" :key="column">
        <button @click="$emit('move', column)" :disabled="!valid" class="board-element">{{ column }}</button>
    </span>
  </div>
</template>

<script>
import axios from 'axios';

export default {
  props: {
    moves: {
      type: Array,
      required: true,
      validator: function (moves) {
        return moves.reduce(function(valid, move) {
          return valid && !isNaN(move) && (move >= 0) && (move <= 6)
        }, true)
      }
    },
    movesAllowed: {
      type: Boolean,
      required: true
    },
    gameResult: {
      type: String,
      required: true,
      validator: function (gameResult) {
        return (gameResult == '1') || (gameResult == '2') || (gameResult == 'tie') || (gameResult == 'incomplete')
      }
    },
    currentPlayerType: {
      type: String,
      required: true,
      validator: function (playerType) {
        return (playerType == 'human') || (playerType == 'computer')
      }
    }
  },
  computed: {
    validMoves: function () {
      if ((this.currentPlayerType == 'human') && (this.movesAllowed) && (this.gameResult == 'incomplete')) {
        let validMoves = []
        for (let candidateMove = 0; candidateMove < 7; candidateMove++) {
            let priorMoveCount = 0
            for (let i = 0; i < this.moves.length; i++){
            if (this.moves[i] == candidateMove) {
                priorMoveCount++
            }
            }
            validMoves.push(priorMoveCount < 6)
        }
        return validMoves
      } else {
          return [false, false, false, false, false, false, false]
      }
    }
  }
}
</script>

<style scoped>
.board-container {
    text-align: center;
}
.board-element {
    display: inline-block;
    width: 30px;
    height: 30px;
}
</style>
