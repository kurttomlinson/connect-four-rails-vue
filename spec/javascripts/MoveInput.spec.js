import { shallow } from 'vue-test-utils'
import Vue from 'vue'
import MoveInput from 'components/MoveInput'

describe('MoveInput.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(MoveInput, {
      propsData: {
        moves: [],
        movesAllowed: true,
        gameResult: 'incomplete',
        currentPlayerType: 'human'
      }
    })
  })

  it('emits a "move" event when a column is clicked', () => {
    const stub = jest.fn()
    wrapper.vm.$on('move', stub)
    for (let i = 0; i < 7; i++) {
      wrapper.findAll('button').at(i).trigger('click')
      expect(stub).toBeCalledWith(i)
    }
  })

  it('disables a button when the column is full', () => {
    wrapper.setProps({
      moves: [0,0,0,0,0,0]
    })
    expect(wrapper.findAll('button').at(0).attributes().disabled).toBe('disabled')
  })

  it('disables all buttons when moves are not allowed', () => {
    wrapper.setProps({
      movesAllowed: false
    })
    for (let i = 0; i < 7; i++) {
      expect(wrapper.findAll('button').at(i).attributes().disabled).toBe('disabled')
    }
  })

  it('disables all buttons when the game is tied', () => {
    wrapper.setProps({
      gameResult: 'tie'
    })
    for (let i = 0; i < 7; i++) {
      expect(wrapper.findAll('button').at(i).attributes().disabled).toBe('disabled')
    }
  })

  it('disables all buttons when the game is won by player 1', () => {
    wrapper.setProps({
      gameResult: '1'
    })
    for (let i = 0; i < 7; i++) {
      expect(wrapper.findAll('button').at(i).attributes().disabled).toBe('disabled')
    }
  })

  it('disables all buttons when the game is won by player 2', () => {
    wrapper.setProps({
      gameResult: '2'
    })
    for (let i = 0; i < 7; i++) {
      expect(wrapper.findAll('button').at(i).attributes().disabled).toBe('disabled')
    }
  })

  it('disables all buttons when the current player is a computer', () => {
    wrapper.setProps({
      currentPlayerType: 'computer'
    })
    for (let i = 0; i < 7; i++) {
      expect(wrapper.findAll('button').at(i).attributes().disabled).toBe('disabled')
    }
  })

  it('enables all buttons when the game is new', () => {
    for (let i = 0; i < 7; i++) {
      expect(wrapper.findAll('button').at(i).attributes().disabled).toBe(undefined)
    }
  })

  it('has the expected html structure', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
