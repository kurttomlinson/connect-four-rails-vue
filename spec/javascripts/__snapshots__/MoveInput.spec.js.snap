// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoveInput.vue has the expected html structure 1`] = `
<div
  class="board-container"
>
  <span>
    <button
      class="board-element"
    >
      0
    </button>
  </span>
  <span>
    <button
      class="board-element"
    >
      1
    </button>
  </span>
  <span>
    <button
      class="board-element"
    >
      2
    </button>
  </span>
  <span>
    <button
      class="board-element"
    >
      3
    </button>
  </span>
  <span>
    <button
      class="board-element"
    >
      4
    </button>
  </span>
  <span>
    <button
      class="board-element"
    >
      5
    </button>
  </span>
  <span>
    <button
      class="board-element"
    >
      6
    </button>
  </span>
</div>
`;
